core/client/src/main/java/alluxio/client/block/stream/NettyPacketWriter.java

```diff
@@ -0,0 +1,446 @@
+ */
+@NotThreadSafe
+public class NettyPacketWriter implements PacketWriter {
+  private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+
+  private static final long PACKET_SIZE =
+      Configuration.getBytes(PropertyKey.USER_NETWORK_NETTY_WRITER_PACKET_SIZE_BYTES);
+  private static final int MAX_PACKETS_IN_FLIGHT =
+      Configuration.getInt(PropertyKey.USER_NETWORK_NETTY_WRITER_BUFFER_SIZE_PACKETS);
+  private static final long WRITE_TIMEOUT_MS =
+      Configuration.getLong(PropertyKey.USER_NETWORK_NETTY_TIMEOUT_MS);
+
+  private final FileSystemContext mContext;
+  private final Channel mChannel;
+  private final InetSocketAddress mAddress;
+  private final long mId;
+  private final long mSessionId;
+  private final Protocol.RequestType mRequestType;
+  private final long mLength;
+
+  private boolean mClosed = false;
+
+  private ReentrantLock mLock = new ReentrantLock();
+  /** The next pos to write to the channel. */
+  @GuardedBy("mLock")
+  private long mPosToWrite = 0;
+  /**
+   * The next pos to queue to the netty buffer. mPosToQueue - mPosToWrite is the data sitting
+   * in the netty buffer.
+   */
+  @GuardedBy("mLock")
+  private long mPosToQueue = 0;
+  @GuardedBy("mLock")
+  private Throwable mPacketWriteException = null;
+  @GuardedBy("mLock")
+  private boolean mDone = false;
+  @GuardedBy("mLock")
+  private boolean mEOFSent = false;
+  /** This condition is met if mPacketWriteException != null or mDone = true. */
+  private Condition mDoneOrFailed = mLock.newCondition();
+  /** This condition is met if mPacketWriteException != null or the buffer is not full. */
+  private Condition mBufferNotFullOrFailed = mLock.newCondition();
+  /** This condition is met if there is nothing in the netty buffer. */
+  private Condition mBufferEmptyOrFailed = mLock.newCondition();
+
+  /**
+   * Creates an instance of {@link NettyPacketWriter}.
+   *
+   * @param context the file system context
+   * @param address the data server network address
+   * @param id the block ID or UFS file ID
+   * @param length the length of the block or file to write, set to Long.MAX_VALUE if unknown
+   * @param sessionId the session ID
+   * @param type the request type (block or UFS file)
+   * @throws IOException it fails to create the object because it fails to acquire a netty channel
+   */
+  public NettyPacketWriter(FileSystemContext context, final InetSocketAddress address, long id,
+      long length, long sessionId, Protocol.RequestType type) throws IOException {
+    mContext = context;
+    mAddress = address;
+    mSessionId = sessionId;
+    mId = id;
+    mRequestType = type;
+    mLength = length;
+    mChannel = context.acquireNettyChannel(address);
+
+    ChannelPipeline pipeline = mChannel.pipeline();
+    if (!(pipeline.last() instanceof RPCMessageDecoder)) {
+      throw new RuntimeException(String.format("Channel pipeline has unexpected handlers %s.",
```

Comment:
NIT: could you put what is expected handler (RPCMessageDecoder?) here in the message too?